### PR TITLE
RB-8 Book model, routes & controllers

### PR DESCRIPTION
**Pull request by vrudnitskaya**
The model for the book was added, as well as all the routes and controllers (CRUD).
Any user, even an unregistered one, can view all books, as well as a specific one.

all-books

single-book

You can also view books added by a specific user.
get-books-userId

The ability to search by one of three parameters or by all at once (author's name, title, ISBN). Books can also be sorted and filtered by different parameters.
books-search

books-sort

Validation, authentication and others errors are handled. 
update-book-invalid-error
update-book-auth-error

### CODE REVIEW

File: src/controllers/bookController.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+    const books = await Book.find(queryObj)
+      .sort(sort)
+      .limit(Number(limit))
+      .skip(Number(skip));
+
+    res.status(StatusCodes.OK).json({ books, count: books.length });
+  } catch (error) {
+    console.error('Error fetching books:', error);
+    next(error);
```
Comment (Tanyaa-a) [inline src/controllers/bookController.js:81]: Great job sending errors to middleware! Should I do it for User as well?

**Review by Tanyaa-a — APPROVED** [review]
You did tremendous work on Books, Valentina!  It's truly impressive

**Comment by Tanyaa-a** [comment]
Everything works good in Postman for Books: search with different params, update the book and delete the book.

File: src/controllers/bookController.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+};
+
+const createBook = async (req, res, next) => {
+  try {
+    const { coverImageUrl, ...bookData } = req.body;
+    bookData.createdBy = req.user.userId;
+    bookData.language = req.body.language || 'English';
+    bookData.coverImageUrl =
+      coverImageUrl || process.env.DEFAULT_COVER_IMAGE_URL;
+
```
Comment (benochi) [inline src/controllers/bookController.js:114]: What happens if there is not a process.env.DEFAULT_COVER_IMAGE_URL amd mno coverImageUrl?

File: src/utils/validateISBN.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+const isbn13Pattern = /^(97[89])(-?\d){10}$/;
+
+const validateISBN10 = (isbn10) => {
+  return !isbn10 || isbn10Pattern.test(isbn10);
+};
+
+const validateISBN13 = (isbn13) => {
+  return !isbn13 || isbn13Pattern.test(isbn13);
+};
+
```
Comment (benochi) [inline src/utils/validateISBN.js:13]: There might be some edge cases where this causes an issue: 
isbn can have dashes: 
Mixed dashes, such as 0-306-40615-2 (valid format).
No dashes at all, such as 0306406152 (also valid).

You're checkig for exactly 9 digits. 

incorrectly groups the first three digits (97[89]) with the next 10 characters, but the dash placement can vary depending on country code, publisher, and item number.
It doesn't correctly handle cases where the dashes divide groups differently:
Example: 978-0-306-40615-7 (valid format but regex fails).
Example: 9780306406157 (valid without dashes but can fail in some scenarios).

File: src/controllers/bookController.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    //General search by "query"
+    if (query) {
+      queryObj.$or = [
+        { title: { $regex: query, $options: 'i' } },
+        { author: { $regex: query, $options: 'i' } },
+        { isbn10: { $regex: query, $options: 'i' } },
+        { isbn13: { $regex: query, $options: 'i' } },
+      ];
+    }
+
```
Comment (benochi) [inline src/controllers/bookController.js:39]: Nothing to fix, I just want you to know that multiple or/and queries are inefficient and you would add indexing in mongodb to fix this.

File: src/middleware/checkObjectId.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const mongoose = require('mongoose');
+
+const checkObjectId = (req, res, next) => {
+  const { id: bookId } = req.params;
+
```
Comment (benochi) [inline src/middleware/checkObjectId.js:6]: checkObjectId creates a new Error object instead of directly passing a BadRequestError or similar custom error.
If that's the intended use that's fine, just a heads up.

File: src/controllers/bookController.js [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+};
+
+const updateBook = async (req, res, next) => {
+  try {
+    const { coverImageUrl, ...updateData } = req.body;
+    const { id: bookId } = req.params;
+    const {
+      user: { userId },
+    } = req;
+
```
Comment (benochi) [inline src/controllers/bookController.js]: we have multiple queries to check the exsistence of the book then to perform the update, which we dont want, mongo has findByIdAndUpdate or findByIdAndDelete
so we only need one API call, 

This is for updateBook and deleteBook

File: src/middleware/error-handler.js [hunk 7]
```diff
@@ -1,33 +1,80 @@
 const { StatusCodes } = require('http-status-codes');
+const { UnauthenticatedError, CustomAPIError } = require('../errors');
+
 const errorHandlerMiddleware = (err, req, res, next) => {
   let customError = {
```
Comment (benochi) [inline src/middleware/error-handler.js:12]: errorHandlerMiddleware exposes the entire stack trace, which could be sensitive information in production, so you could modify this based on development/production in the real world.  Just FYI.

File: src/controllers/bookController.js [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+      .status(StatusCodes.CREATED)
+      .json({ msg: 'The book has been successfully created.', book });
+  } catch (error) {
+    next(error);
+  }
+};
+
+const updateBook = async (req, res, next) => {
+  try {
+    const { coverImageUrl, ...updateData } = req.body;
```
Comment (benochi) [inline src/controllers/bookController.js]: We might not want them to be able to update every field, IE: createdBy or other immutable categories. Not a big deal, but FYI.

File: src/controllers/bookController.js [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    const book = await Book.findById(bookId);
+    if (!book) {
+      return next(new NotFoundError(`No book found with id: ${bookId}`));
+    }
+
+    if (book.createdBy.toString() !== userId) {
+      return next(new BadRequestError('You are not authorized to delete this book'));
+    }
+
+    await Book.findOneAndDelete({ _id: bookId, createdBy: userId });
```
Comment (benochi) [inline src/controllers/bookController.js]: We don't need created by here, since the bookId will be uique and is fectched above(const book = await Book.findById(bookId); you can use the book._id 

*but* we should just use findbyIDandDelete and only make one API call.

File: src/app.js [hunk 10]
```diff
@@ -21,5 +23,16 @@ app.use(favicon(__dirname + '/public/favicon.ico'));
```
Comment (benochi) [inline src/app.js:25]: nitpick but these should be flipped and just serve from public:
app.use(favicon(__dirname + '/public/favicon.ico'));
app.use('/public', express.static('public'));

File: src/middleware/authentication.js [hunk 11]
```diff
@@ -5,7 +5,7 @@ const auth = async (req, res, next) => {
   const authHeader = req.headers.authorization;
 
   if (!authHeader || !authHeader.startsWith('Bearer ')) {
-    throw new UnauthenticatedError('Authentication invalid');
+    return next(new UnauthenticatedError('Authentication invalid'));
```
Comment (benochi) [inline src/middleware/authentication.js:8]: We might want to add a expiresIn: '2h' to JWTs, if so, we can check for expiration like so. 

const jwt = require('jsonwebtoken');
const { UnauthenticatedError } = require('../errors');

const auth = async (req, res, next) => {
  const authHeader = req.headers.authorization;

  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return next(new UnauthenticatedError('Authentication invalid'));
  }

  const token = authHeader.split(' ')[1];
  try {
    const payload = jwt.verify(token, process.env.JWT_SECRET);
    req.user = { userId: payload.userId, firstName: payload.firstName };
    next();
  } catch (error) {
    if (error.name === 'TokenExpiredError') {
      return next(new UnauthenticatedError('Token has expired. Please log in again.'));
    }
    return next(new UnauthenticatedError('Authentication invalid'));
  }
};

module.exports = auth;

Comment (benochi) [reply src/middleware/authentication.js:8]: and this where we are creating them: 
const jwt = require('jsonwebtoken');

const createJWT = (user) => {
  return jwt.sign(
    { userId: user._id, firstName: user.firstName },
    process.env.JWT_SECRET,
    { expiresIn: '2h' } 
  );
};

File: src/controllers/bookController.js [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    next(error);
+  }
+};
+
+const createBook = async (req, res, next) => {
+  try {
+    const { coverImageUrl, ...bookData } = req.body;
+    bookData.createdBy = req.user.userId;
+    bookData.language = req.body.language || 'English';
+    bookData.coverImageUrl =
```
Comment (benochi) [inline src/controllers/bookController.js:112]: coverImageUrl logic in createBook and updateBook does not validate the integrity of external URLs. just a heads up.

Comment (benochi) [reply src/controllers/bookController.js:112]: solution would be to: 
Use a Validation Utility: Create a utility function to validate the integrity of the coverImageUrl. This function should:

Ensure the URL has a valid structure.
Check the file extension to ensure it points to an image.
Integrate Validation in createBook and updateBook: Use the validation function before accepting the coverImageUrl. If the validation fails, return an error response.

You can do this if there is time at the end.

**Review by benochi — APPROVED** [review]
Please read notes, most are just FYIs.

File: src/controllers/bookController.js (hunk 2 above)
Comment (vrudnitskaya) [reply src/controllers/bookController.js:114]: I will fix this when I integrate uploading covers to Cloudinary. I’ll upload a default image there.

File: src/app.js (hunk 10 above)
Comment (vrudnitskaya) [reply src/app.js:23]: like this?
```
app.use('/public', express.static('public'));
app.use(express.static('public')); 
```

File: src/middleware/authentication.js (hunk 11 above)
Comment (vrudnitskaya) [reply src/middleware/authentication.js:8]: We've already have this in the user model, so I will just change the token lifetime in `.env`
